src/cmap/auth/mongodb_oidc.ts

```diff
@@ -1,39 +1,136 @@
-/**
- * TODO: NODE-5035: Make API public
- *
- * @internal */
+import { MongoInvalidArgumentError, MongoMissingCredentialsError } from '../../error';
+import type { Callback } from '../../utils';
+import type { HandshakeDocument } from '../connect';
+import { type AuthContext, AuthProvider } from './auth_provider';
+import type { MongoCredentials } from './mongo_credentials';
+import { AwsDeviceWorkflow } from './mongodb_oidc/aws_device_workflow';
+import { CallbackWorkflow } from './mongodb_oidc/callback_workflow';
+import type { Workflow } from './mongodb_oidc/workflow';
+
+/** @public */
 export interface OIDCMechanismServerStep1 {
-  authorizeEndpoint?: string;
+  authorizationEndpoint?: string;
   tokenEndpoint?: string;
-  deviceAuthorizeEndpoint?: string;
+  deviceAuthorizationEndpoint?: string;
   clientId: string;
   clientSecret?: string;
   requestScopes?: string[];
 }
 
-/**
- * TODO: NODE-5035: Make API public
- *
- * @internal */
+/** @public */
 export interface OIDCRequestTokenResult {
   accessToken: string;
   expiresInSeconds?: number;
   refreshToken?: string;
 }
 
-/**
- * TODO: NODE-5035: Make API public
- *
- * @internal */
+/** @public */
 export type OIDCRequestFunction = (
-  idl: OIDCMechanismServerStep1
+  principalName: string,
+  serverResult: OIDCMechanismServerStep1,
+  timeout: AbortSignal | number
 ) => Promise<OIDCRequestTokenResult>;
 
-/**
- * TODO: NODE-5035: Make API public
- *
- * @internal */
+/** @public */
 export type OIDCRefreshFunction = (
-  idl: OIDCMechanismServerStep1,
-  result: OIDCRequestTokenResult
+  principalName: string,
+  serverResult: OIDCMechanismServerStep1,
+  result: OIDCRequestTokenResult,
+  timeout: AbortSignal | number
 ) => Promise<OIDCRequestTokenResult>;
+
+/** @internal */
+export const OIDC_WORKFLOWS = {
+  callback: new CallbackWorkflow(),
+  aws: new AwsDeviceWorkflow(),
+  azure: undefined,
+  gcp: undefined,
+  // Set the protype to null to avoid false returns and keep types.
+  __proto__: null
+};
+
+/**
+ * OIDC auth provider.
+ */
+export class MongoDBOIDC extends AuthProvider {
+  /**
+   * Instantiate the auth provider.
+   */
+  constructor() {
+    super();
+  }
+
+  /**
+   * Authenticate using OIDC
+   */
+  override auth(authContext: AuthContext, callback: Callback): void {
+    const { connection, credentials, response } = authContext;
+
+    if (response && response.speculativeAuthenticate) {
+      return callback();
+    }
+
+    if (!credentials) {
+      return callback(new MongoMissingCredentialsError('AuthContext must provide credentials.'));
+    }
+
+    getWorkflow(credentials, (error, workflow) => {
+      if (error || !workflow) {
```

Comment:
same comment here about `!workflow` being a runtime error as in L107

Reply:
Changed to pass a `MongoRuntimeError`